src/java.base/share/classes/java/lang/foreign/GroupLayout.java

```diff
@@ -73,4 +76,136 @@ public sealed interface GroupLayout extends MemoryLayout permits StructLayout, U
+     * <p>
+     * Unnamed elements in this group will be ignored.
+     * Unmatched elements (with respect to the name) in this group layout will be ignored.
+     * <p>
+     * The returned function will respect the byte orderings and alignment constraints of this
+     * group layout.
+     * <p>
+     * The returned Function may throw an {@link IllegalArgumentException} if it, for any reason, fails
+     * to extract a {@link Record}. An example of such a failure is if the applied memory segment is too
+     * small for the layout at hand.
+     * <p>
+     * The example below shows how to extract an instance of a public {@code Point} record class
+     * from a {@link MemorySegment}:
+     * {@snippet lang = java:
+     *     MemorySegment segment = MemorySegment.ofArray(new int[]{3, 4});
+     *
+     *     public record Point(int x, int y){}
+     *
+     *     var pointLayout = MemoryLayout.structLayout(
+     *         JAVA_INT.withName("x"),
+     *         JAVA_INT.withName("y")
+     *     );
+     *
+     *     Function<MemorySegment, Point> pointExtractor = pointLayout.recordMapper(Point.class);
+     *
+     *     // Extracts a new Point from the provided MemorySegment
+     *     Point point = pointExtractor.apply(segment); // Point[x=3, y=4]
+     * }
+     * <p>
+     * Boxing, widening and narrowing must be explicitly handled by user code. In the following example, the above
+     * {@code Point} (using primitive {@code int x} and {@code int y} coordinates) are explicitly mapped to
+     * a narrowed point type (instead using primitive {@code byte x} and {@code byte y} coordinates):
+     * <p>
+     * {@snippet lang = java:
+     *     public record NarrowedPoint(byte x, byte y) {
+     *
+     *         static NarrowedPoint fromPoint(Point p) {
+     *             return new NarrowedPoint((byte) p.x, (byte) p.y);
+     *         }
+     *     }
+     *
+     *     Function<MemorySegment, NarrowedPoint> narrowedPointExtractor =
+     *             pointLayout.recordMapper(Point.class)
+     *                     .andThen(NarrowedPoint::fromPoint);
+     *
+     *     // Extracts a new NarrowedPoint from the provided MemorySegment
+     *     NarrowedPoint narrowedPoint = narrowedPointExtractor.apply(segment); // NarrowedPoint[x=3, y=4]
+     * }
+     *
+     * @param <R> record type
+     * @param type the type (Class) of the record
+     * @throws IllegalArgumentException if the provided record {@code type} contains components for which
+     *                                  there are no exact mapping (of names and types) in this group layout
+     *                                  or if the provided {@code type} is not public or if the method is otherwise
+     *                                  unable to create a record mapper as specified above.
+     * @since 21
+     */
+    <R extends Record> Function<MemorySegment, R> recordMapper(Class<R> type);
```

Comment:
you also need to throw an IAE if type is Record.class, java.lang.Record is the only subtype of java.lang.Record which is not a record

Reply:
I wonder if an overload which takes a MethodHandles.Lookup might help with the accessibility issues (but we can add that later)

Reply:
Yes ! and i'm not even sure that having a version that does not take a Lookup make sense.